x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/aggregation/RollupDataExtractorFactory.java

```diff
@@ -0,0 +1,219 @@
+            );
+            return;
+        }
+
+        listener.onResponse(new RollupDataExtractorFactory(client, datafeed, job));
+    }
+
+    private static boolean validInterval(long datafeedInterval, ParsedRollupCaps rollupJobGroupConfig) {
+        if (rollupJobGroupConfig.hasDatehistogram() == false) {
+            return false;
+        }
+        if (DateTimeZone.UTC.toString().equalsIgnoreCase(rollupJobGroupConfig.getTimezone()) == false) {
+            return false;
+        }
+        try {
+            long jobInterval = validateAndGetCalendarInterval(rollupJobGroupConfig.getInterval());
+            return datafeedInterval % jobInterval == 0;
+        } catch (ElasticsearchStatusException exception) {
+            return false;
+        }
+    }
+
+    private static void flattenAggregations(final Collection<AggregationBuilder> datafeedAggregations,
+                                            final AggregationBuilder datafeedHistogramAggregation,
+                                            final List<AggregationBuilder> flattenedAggregations) {
+        for (AggregationBuilder aggregationBuilder : datafeedAggregations) {
+            if (aggregationBuilder.equals(datafeedHistogramAggregation) == false) {
+                flattenedAggregations.add(aggregationBuilder);
+            }
+            flattenAggregations(aggregationBuilder.getSubAggregations(), datafeedHistogramAggregation, flattenedAggregations);
+        }
+    }
+
+    private static boolean hasAggregations(ParsedRollupCaps rollupCaps, List<AggregationBuilder> datafeedAggregations) {
+        for (AggregationBuilder aggregationBuilder : datafeedAggregations) {
+            String type = aggregationBuilder.getType();
+            String field = ((ValuesSourceAggregationBuilder) aggregationBuilder).field();
+            if (aggregationBuilder instanceof TermsAggregationBuilder) {
+                if (rollupCaps.supportedTerms.contains(field) == false) {
+                    return false;
+                }
+            } else {
+                if (rollupCaps.supportedMetrics.contains(field + "_" + type) == false) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+
+    private static class ParsedRollupCaps {
+        private final Set<String> supportedMetrics;
+        private final Set<String> supportedTerms;
+        private final Map<String, Object> datehistogramAgg;
+        private static List<String> aggsToIgnore = Arrays.asList(HistogramAggregationBuilder.NAME, DateHistogramAggregationBuilder.NAME);
+
+
+        private static ParsedRollupCaps fromJobFieldCaps(Map<String, RollupFieldCaps> rollupFieldCaps, String timeField) {
+            Map<String, Object> datehistogram = null;
+            RollupFieldCaps timeFieldCaps = rollupFieldCaps.get(timeField);
+            if ((timeFieldCaps == null) == false) {
```

Comment:
simplify: `if (timeFieldCaps != null)`?

Reply:
I am fine with that simplification (it is my preference to use `!=` other than `== false`) but the prevailing pattern I have seen throughout elastic is `== false`.

Reply:
`{expression} == false` is preferred instead of `!{expression}`. Other than that, `!=` is used normally.